src/hb/shape_wasm.rs

```diff
@@ -0,0 +1,566 @@
+use alloc::{borrow::ToOwned, ffi::CString, format};
+use core::ffi::CStr;
+use ttf_parser::{GlyphId, Tag};
+use wasmi::{self, AsContextMut, Caller, Engine, Linker, Module, Store};
+
+use super::{
+    buffer::{hb_buffer_t, GlyphPosition},
+    face::hb_glyph_extents_t,
+    hb_font_t, hb_glyph_info_t,
+    ot_shape::{hb_ot_shape_context_t, shape_internal},
+    ot_shape_plan::hb_ot_shape_plan_t,
+};
+
+struct ShapingData<'a> {
+    font: &'a hb_font_t<'a>,
+    plan: &'a hb_ot_shape_plan_t,
+    buffer: &'a mut hb_buffer_t,
+}
+
+pub(crate) fn shape_with_wasm(
+    font: &hb_font_t,
+    plan: &hb_ot_shape_plan_t,
+    buffer: &mut hb_buffer_t,
+) -> Option<()> {
+    // If font has no Wasm blob just return None to carry on as usual.
+    let wasm_blob = font
+        .raw_face()
+        .table(ttf_parser::Tag::from_bytes(b"Wasm"))?;
+
+    let data = ShapingData { font, plan, buffer };
+
+    let mut store = Store::new(&Engine::default(), data);
```

Comment:
If the Wasm binaries are rather big and only parts of them are usually used (not all functions defined in them), then it might make sense to use [Wasmi's lazy compilation feature](https://docs.rs/wasmi/0.34.0/wasmi/struct.Config.html#method.compilation_mode) in order to improve startup times significantly. Expected speed-up for larger Wasm blobs is ~10-40x.

You can do this as follows:

```rust
let mut config = wasmi::Config::default();
engine.compilation_mode(wasmi::CompilationMode::Lazy);
let engine = wasmi::Engine::new(&config);

// now you can use engine
```

Reply:
Thanks for the tip.